pkg/controllers/updaterun/validation.go

```diff
@@ -0,0 +1,319 @@
+		klog.ErrorS(failedErr, "The stage has failed", "stageCond", stageSucceedCond, "clusterStagedUpdateRun", klog.KObj(updateRun))
+		return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, failedErr.Error())
+	} else if stageStartedCond != nil {
+		// The stage is still updating.
+		if updatingStageIndex != -1 {
+			// There should be only one stage updating at a time.
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the stage `%s` is updating, but there is already a stage with index `%d` updating", stageStatus.StageName, updatingStageIndex))
+			klog.ErrorS(unexpectedErr, "Detected more than one updating stages", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		if curStage != lastFinishedStageIndex+1 {
+			// The current updating stage is not right after the last finished stage.
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the updating stage `%s` is not right after the last finished stage with index `%d`", stageStatus.StageName, lastFinishedStageIndex))
+			klog.ErrorS(unexpectedErr, "There's not yet started stage before the updating stage", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		updatingStageIndex = curStage
+		// Collect the updating clusters.
+		var updatingClusters []string
+		for j := range stageStatus.Clusters {
+			clusterStartedCond := meta.FindStatusCondition(stageStatus.Clusters[j].Conditions, string(placementv1alpha1.ClusterUpdatingConditionStarted))
+			clusterFinishedCond := meta.FindStatusCondition(stageStatus.Clusters[j].Conditions, string(placementv1alpha1.ClusterUpdatingConditionSucceeded))
+			if condition.IsConditionStatusTrue(clusterStartedCond, updateRun.Generation) &&
+				!(condition.IsConditionStatusTrue(clusterFinishedCond, updateRun.Generation) || condition.IsConditionStatusFalse(clusterFinishedCond, updateRun.Generation)) {
+				updatingClusters = append(updatingClusters, stageStatus.Clusters[j].ClusterName)
+			}
+		}
+		// We don't allow more than one clusters to be updating at the same time.
+		if len(updatingClusters) > 1 {
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("more than one cluster is updating in the stage `%s`, clusters: %v", stageStatus.StageName, updatingClusters))
+			klog.ErrorS(unexpectedErr, "Detected more than one updating clusters in the stage", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+	}
+	return updatingStageIndex, lastFinishedStageIndex, nil
+}
+
+// validateDeleteStageStatus validates the delete stage in the clusterStagedUpdateRun.
+// It returns the updating stage index, or any error encountered.
+func validateDeleteStageStatus(
```

Comment:
The clusters in the deletionStage are not validated. I think it's ok because cluster might unjoin the fleet?

Reply:
Do you mean that we don't compare it with the exisitng list?  I think we still should but since the deleted binding will disapper, it means the new list should be a subset of the old?